### DISCUSSION TITLE
Use Mantine (React components library)

### DISCUSSION THREAD

**Discussion by karlssten** [eclipse-theia/theia#13041] [Q&A]
Hi! I would like to use [Mantine](https://github.com/mantinedev/mantine) components when writing widgets. Has anyone done this before? What's the recommended approach?

According to Mantine's [docs](https://mantine.dev/getting-started/#get-started-without-framework), it requires wrapping the application with ``MantineProvider`` like so:

```javascript
import { MantineProvider, createTheme } from '@mantine/core';

const theme = createTheme({
  /** Put your mantine theme override here */
});

function Demo() {
  return (
    <MantineProvider theme={theme}>
      <App />
    </MantineProvider>
  );
}
```

Since Theia does not have a traditional single 'entry point', how should it be done? Should it be added in the main [FrontendApplication](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/frontend-application.ts) or the [ApplicationShell](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/shell/application-shell.ts)? Is this a good approach?

On the other hand, Mantine's styles can be customized by overriding its theme like this:

```javascript
import { MantineProvider, createTheme, rem } from '@mantine/core';

const theme = createTheme({
  colors: {
    // Add your color
    deepBlue: ['#E9EDFC', '#C1CCF6', '#99ABF0' /* ... */],
    // or replace default theme color
    blue: ['#E9EDFC', '#C1CCF6', '#99ABF0' /* ... */],
  },

  shadows: {
    md: '1px 1px 3px rgba(0, 0, 0, .25)',
    xl: '5px 5px 3px rgba(0, 0, 0, .25)',
  },

  headings: {
    fontFamily: 'Roboto, sans-serif',
    sizes: {
      h1: { fontSize: rem(36) },
    },
  },
});

function Demo() {
  return (
    <MantineProvider theme={theme}>
      <App />
    </MantineProvider>
  );
}
```

This could be done by injecting the ``ThemeService`` and using it's ``getCurrentTheme()`` method, right? In what part of the application lifecycle should it be done?

I don't have much experience with Theia nor Mantine, so any information or opinion is highly appreciated, thank you!

**Comment by msujew** [comment]
Hey @fermar221,

since neither the `FrontendApplication` nor the `ApplicationShell` actually run React (they run on pure HTML), there would be nowhere to inject the `MantineProvider` in those classes.

I assume the recommended way would be to just wrap your custom react widgets into the provider:

```ts
export class MyWidget extends ReactWidget {
  protected render(): JSX.Element {
    return <MantineProvider theme={theme}>
      <... />
    </MantineProvider>
  }
}
```

**Reply by karlssten** [reply]
Hi @msujew, thank you for replying!

That would work, yes. The issue is that Mantine expects ``MantineProvider`` to sit on top, in their [docs ](https://mantine.dev/theming/mantine-provider/) they say that "_it must be rendered at the root of your application and should be used only once_.", which is why I was hoping to find a way of integrating it above the widget level. I'm not sure what the overhead for having it in every widget is, or how scalable that would be.

Is running React in ``FrontendApplication`` or ``ApplicationShell`` not a good idea/not doable?

**Reply by msujew** [reply]
> Is running React in FrontendApplication or ApplicationShell not a good idea/not doable?

Both `ApplicationShell` (and to a more minor extend `FrontendApplication`) are just managing the layout of other widgets (i.e. navigator, editors, terminals, etc.). They don't perform any actual rendering but work on HTML elements directly. I don't think it is possible to work React into this.

While Theia makes use of a lot of React, it's not a React app in itself. It just uses React portalling to allow to render widgets (i.e. individual elements of the UI) using React.
